### PR TITLE
Stats leaderboards

### PR DESCRIPTION
**Pull request by Vexrax**
# Implements leaderboard on the stats page: 
image
image

### Notes
- Risen / ALL tournmant options will not show leaderboards to save the DB
- Used n >= 4 for the min amount of games, lmk if it should be higher 
- Requires the "Improve Stats Caching" as a followup so that we ingest games when they end

### CODE REVIEW

File: Common/utils.ts [hunk 1]
```diff
@@ -220,4 +221,187 @@ export function ObjectArrayToCsv(objects: { [key: string]: any }[], headers?: st
   ));
 
   return [dictionaryKeys.join(','), ...dictValuesAsCsv].join('\n');
+}
+
+export function combine(object1: PlayerStatModel, object2: PlayerStatModel): PlayerStatModel {
```
Comment (Vexrax) [inline Common/utils.ts:226]: Ik this function is duplicating code, but i will refactor when i do the next task

Comment (afieldi) [reply Common/utils.ts:226]: Ok

File: front-end/src/api/leaderboards.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import {PlayerChampionStatsRequest} from "../../../Common/Interface/Internal/player";
+import {GameRoles} from "../../../Common/Interface/General/gameEnums";
+import {MakeBackendCall} from "./_call";
+import {GetLeaderboardResponse} from "../../../Common/Interface/Internal/leaderboard";
+import PlayerStatModel from "../../../Common/models/playerstat.model";
+import {combine} from "../../../Common/utils";
+
+export async function getFlattenedLeaderboard(seasonId?: number, risenOnly?: boolean, roleId?: string) : Promise<PlayerStatModel[]> {
+    console.log(`getting leaderboard for SID: ${seasonId} RISEN: ${risenOnly} ROLE: ${roleId}`)
```
Comment (afieldi) [inline front-end/src/api/leaderboards.ts:9]: When you refactor replace this with the logger

File: front-end/src/common/utils.ts [hunk 3]
```diff
@@ -33,4 +33,12 @@ export function SaveBlob(blob: any, fileName: string) {
 
   // Clean up and remove the link
   link.parentNode?.removeChild(link);
-}
\ No newline at end of file
+}
+
+export function nth(n: number): string {
+  return["st","nd","rd"][((n+90)%100-10)%10-1]||"th"
```
Comment (afieldi) [inline front-end/src/common/utils.ts:39]: KEKW

File: front-end/src/components/player-page/header.tsx [hunk 4]
```diff
@@ -11,33 +11,35 @@ import LoadingButton from "../loading-button/LoadingButton";
 
 interface Props {
   playerOverview: PlayerOverviewResponse | undefined;
+  onUpdate: () => void;
 }
 
-export default function PlayerPageHeader({playerOverview}: Props)
+export default function PlayerPageHeader(playerOverviewProps: Props)
```
Comment (afieldi) [inline front-end/src/components/player-page/header.tsx:17]: Bruh. Why

File: front-end/src/components/player-page/stats/leaderboard-ranking.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import React from "react";
+import {getNumberWithSuffix} from "../../../common/utils";
+
+interface LeaderboardRankingProps {
+    rank: number
+    leagueAvg: number
+    totalPLayersOnLeaderboard: number
+}
+
+const FIRST = "#FFA500";
```
Comment (afieldi) [inline front-end/src/components/player-page/stats/leaderboard-ranking.tsx:11]: Refactor these out into a more global file. Tbh this should be part of the theme.

File: front-end/src/components/player-page/stats/leaderboard-ranking.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+const SECOND = "#85BCBF";
+const THIRD = "#834B24";
+const NTH = "#565554"
+
+const rankToColorMap: Record<number, string> = {
```
Comment (afieldi) [inline front-end/src/components/player-page/stats/leaderboard-ranking.tsx:16]: Tbh just make this a function instead.
```
getColor(rank) {
  switch(rank) {
    case 1:
       return theme.x.first.main;
    .....
  }
```

Comment (afieldi) [reply front-end/src/components/player-page/stats/leaderboard-ranking.tsx:16]: If you want. I don't think its too big of a diff. Might just be easier to look at.

File: front-end/src/components/player-page/stats/stats-generators/AverageVisionScoreStatGenerator.ts [hunk 7]
```diff
@@ -2,21 +2,15 @@ import {BaseStatGenerator} from "./BaseStatsGenerator";
...
-
     getStatTitle(): string {
         return "AVG VS"
     }
 
     getToolTip(): string {
         return "Average Vision Score Per Game";
     }
+
+    getStatValue(playerStatsModel: PlayerStatModel): number {
```
Comment (afieldi) [inline front-end/src/components/player-page/stats/stats-generators/AverageVisionScoreStatGenerator.ts:13]: Add rounding

File: front-end/src/components/player-page/stats/stats-generators/DPMStatGenerator.ts [hunk 8]
```diff
@@ -2,23 +2,16 @@ import {BaseStatGenerator} from "./BaseStatsGenerator";
...
     getStatTitle(): string {
         return "DPM"
     }
 
     getToolTip(): string {
         return "Damage Per Min";
     }
+
+    getStatValue(playerStatsModel: PlayerStatModel): number {
+        let minutes = playerStatsModel.gameLength / 60
```
Comment (afieldi) [inline front-end/src/components/player-page/stats/stats-generators/DPMStatGenerator.ts:14]: ig this one is probably ok, but maybe refactor to use the util function for riot gamelength to minutes. So incase for some reason this changes.

File: front-end/src/pages/player/player.tsx [hunk 9]
```diff
@@ -75,6 +76,30 @@ function PlayerPage()
     }
   }
 
+  async function loadLeaderboards() {
+    if (seasonId === "ALL" || seasonId === "RISEN") {
```
Comment (afieldi) [inline front-end/src/pages/player/player.tsx:80]: Thank fuck

File: front-end/src/components/player-page/header.tsx (hunk 4 above)
Comment (Vexrax) [reply front-end/src/components/player-page/header.tsx:17]: I had to update the props to pass an update function so this needed to be changed

File: front-end/src/components/player-page/stats/leaderboard-ranking.tsx (hunk 6 above)
Comment (Vexrax) [reply front-end/src/components/player-page/stats/leaderboard-ranking.tsx:16]: Personally i find maps easier to look at so im going to keep it the same

File: front-end/src/components/player-page/stats/stats-generators/AverageVisionScoreStatGenerator.ts (hunk 7 above)
Comment (Vexrax) [reply front-end/src/components/player-page/stats/stats-generators/AverageVisionScoreStatGenerator.ts:13]: We dont want rounding here, this function is to get the raw stat value. Rounding should only happen when the value gets converted to a string. The raw value is useful for leaderboard rankings where you care about decimal precision.

File: front-end/src/api/leaderboards.ts (hunk 2 above)
Comment (Vexrax) [reply front-end/src/api/leaderboards.ts:9]: We dont have a logger object in the frontend package>

File: front-end/src/common/utils.ts [hunk 10]
```diff
@@ -33,4 +33,16 @@ export function SaveBlob(blob: any, fileName: string) {
...
+}
+
+export function getNumberWithSuffix(n: number): string {
+  return `${n}${nth(n)}`
+}
+
+export function getMinutesFromRiotTime(riotTime: number): number {
```
Comment (afieldi) [inline front-end/src/common/utils.ts:46]: I hate to do this to you. But https://github.com/afieldi/RisenStats/blob/master/Common/utils.ts this already exists.

Comment (Vexrax) [reply front-end/src/common/utils.ts]: PAIN

**Review by afieldi — APPROVED** [review]

**Pull request merged**